### ISSUE TITLE
Possible issue with Wx::AUI::AuiNotebook

### ISSUE THREAD

**Issue by Randalphwa** [mcorino/wxRuby3#141]
I'm not sure if this is a problem with my code, or a problem with wxRuby3. I first created a `Wx::Notebook` with a single page which works fine. I then changed that to a `Wx::AUI::AuiNotebook`, and added a `require 'wx/aui'` line. The dialog displays fine, but when I terminate the dialog, Ruby crashes:

```
 *  Executing task: ruby rb_main.rb 

D:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt/lib/wx/core/app.rb:43: [BUG] Segmentation fault
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [x64-mingw-ucrt]

-- Control frame information -----------------------------------------------
c:0006 p:---- s:0024 e:000023 CFUNC  :get_client_area_origin
c:0005 p:---- s:0021 e:000020 CFUNC  :main_loop
c:0004 p:0044 s:0017 e:000016 METHOD D:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt/lib/wx/core/app.rb:43
c:0003 p:0029 s:0012 e:000011 METHOD D:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt/lib/wx/core/app.rb:59
c:0002 p:0056 s:0006 e:000005 EVAL   rb_main.rb:105 [FINISH]
c:0001 p:0000 s:0003 E:000a90 DUMMY  [FINISH]

-- Ruby level backtrace information ----------------------------------------
rb_main.rb:105:in `<main>'
D:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt/lib/wx/core/app.rb:59:in `run'
D:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt/lib/wx/core/app.rb:43:in `run'
D:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt/lib/wx/core/app.rb:43:in `main_loop'
D:/Ruby32-x64/lib/ruby/gems/3.2.0/gems/wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt/lib/wx/core/app.rb:43:in `get_client_area_origin'

-- C level backtrace information -------------------------------------------
C:\windows\SYSTEM32\ntdll.dll(NtWaitForSingleObject+0x14) [0x00007ffb29a0edd4]
C:\windows\System32\KERNELBASE.dll(WaitForSingleObjectEx+0x8e) [0x00007ffb27143f8e]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_vm_bugreport+0x256) [0x00007fface942ee6]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_bug_for_fatal_signal+0x80) [0x00007fface7405b0]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_shape_memsize+0x542) [0x00007fface88e502]
 [0x00007ff70d1a1e22]
C:\windows\System32\ucrtbase.dll(_C_specific_handler+0xa0) [0x00007ffb26e57f30]
C:\windows\SYSTEM32\ntdll.dll(_chkstk+0x12f) [0x00007ffb29a13dff]
C:\windows\SYSTEM32\ntdll.dll(RtlFindCharInUnicodeString+0xa96) [0x00007ffb2998e456]
C:\windows\SYSTEM32\ntdll.dll(KiUserExceptionDispatcher+0x2e) [0x00007ffb29a12dee]
 [0x000000000000302e]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\lib\wxruby_core.so(Z16GC_mark_wxWindowPv+0x125) [0x00007ffacdbaeea5]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\lib\wxruby_aui.so(Init_wxAuiTabArt+0xac69) [0x00007ffad1be5979]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_gc_verify_internal_consistency+0xaa8) [0x00007fface76e798]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_gc_verify_internal_consistency+0x1a88) [0x00007fface76f778]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_xcalloc_mul_add_mul+0xe6a) [0x00007fface77356a]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_wb_protected_newobj_of+0x7c) [0x00007fface7742bc]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_str_unlocktmp+0x4fc8) [0x00007fface8a6e68]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\lib\wxruby_core.so(Init_wxTopLevelWindow+0x5440) [0x00007ffacd673d00]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_vm_invoke_proc+0x1eb) [0x00007fface92b73b]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_eval_cmd_kw+0x5a3) [0x00007fface930623]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_funcallv+0x11) [0x00007fface930871]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_protect+0x1a4) [0x00007fface74bfe4]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\lib\wxruby_core.so(Z14wxRuby_FuncallRbyyy+0x18a36) [0x00007ffacdbccbd6]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\lib\wxruby_core.so(Z14wxRuby_Funcallyyiz+0x96) [0x00007ffacdbb4046]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\lib\wxruby_core.so(Init_wxTopLevelWindow+0x903e) [0x00007ffacd6778fe]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\ext\wxmsw32u_core_gcc_custom.dll(ZNK20wxTopLevelWindowBase16DoClientToScreenEPiS0_+0x19) [0x00007ffac89d2289]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\ext\wxmsw32u_core_gcc_custom.dll(ZN8wxWindow14InitMouseEventER12wxMouseEventiij+0x10f) [0x00007ffac87b54cf]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\ext\wxmsw32u_core_gcc_custom.dll(ZN8wxWindow16HandleMouseEventEjiij+0x7e) [0x00007ffac87b558e]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\ext\wxmsw32u_core_gcc_custom.dll(ZN8wxWindow15HandleMouseMoveEiij+0x18e) [0x00007ffac87ba35e]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\ext\wxmsw32u_core_gcc_custom.dll(ZN8wxWindow16MSWHandleMessageEPxjyx+0x6b9) [0x00007ffac87c23f9]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\ext\wxmsw32u_core_gcc_custom.dll(ZN8wxWindow13MSWWindowProcEjyx+0x2c) [0x00007ffac87af0bc]
C:\windows\System32\USER32.dll(DispatchMessageW+0x741) [0x00007ffb28c98211]
C:\windows\System32\USER32.dll(DispatchMessageW+0x201) [0x00007ffb28c97cd1]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\ext\wxmsw32u_core_gcc_custom.dll(ZN14wxGUIEventLoop8DispatchEv+0x159) [0x00007ffac87e3409]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\ext\wxbase32u_gcc_custom.dll(ZN17wxEventLoopManual13ProcessEventsEv+0x33) [0x00007ffad5aa56c3]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\ext\wxbase32u_gcc_custom.dll(ZN17wxEventLoopManual5DoRunEv+0x108) [0x00007ffad5aa57e8]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\ext\wxbase32u_gcc_custom.dll(ZN15wxEventLoopBase3RunEv+0x58) [0x00007ffad5aa54a8]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\ext\wxbase32u_gcc_custom.dll(ZN16wxAppConsoleBase8MainLoopEv+0x70) [0x00007ffad5a75820]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\ext\wxbase32u_gcc_custom.dll(Z12wxInitializeRiPPc+0xb1) [0x00007ffad5ae9651]
D:\Ruby32-x64\lib\ruby\gems\3.2.0\gems\wxruby3-0.9.0.pre.rc.3-x64-mingw-ucrt\lib\wxruby_core.so(Init_wxGUIEventLoop+0x13f9) [0x00007ffacd5fd8f9]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_error_arity+0x147) [0x00007fface918b27]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_vm_opt_newarray_max+0x3ec) [0x00007fface9344cc]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_vm_exec+0x65c) [0x00007fface925b4c]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(rb_call_end_proc+0x130) [0x00007fface7451f0]
D:\Ruby32-x64\bin\x64-ucrt-ruby320.dll(ruby_run_node+0xa5) [0x00007fface74afd5]
 [0x00007ff70d1a2a20]
 [0x00007ff70d1a12ee]
 [0x00007ff70d1a1406]
C:\windows\System32\KERNEL32.DLL(BaseThreadInitThunk+0x1d) [0x00007ffb27a626ad]
```
Invoking the dialog is done via:

```Ruby
def on_ruby_test_dlg
  test_dlg = RubyTest.new(self)
  test_dlg.show_modal
  test_dlg.destroy
end
```

The dialog itself is:

```Ruby

WX_GLOBAL_CONSTANTS = true unless defined? WX_GLOBAL_CONSTANTS

require 'wx/core'
require 'wx/aui'

class RubyTest < Wx::Dialog
  def initialize(parent, id=Wx::ID_ANY, title="Ruby Test Dialog",
        pos=Wx::DEFAULT_POSITION, size=Wx::DEFAULT_SIZE,
        style=Wx::DEFAULT_DIALOG_STYLE)

    super(parent, id, title, pos, size, style)

    dlg_sizer = Wx::BoxSizer.new(Wx::VERTICAL)

    box_sizer_2 = Wx::BoxSizer.new(Wx::HORIZONTAL)

    @notebook = Wx::AUI::AuiNotebook.new(self, Wx::ID_ANY,
      Wx::DEFAULT_POSITION, Wx::DEFAULT_SIZE, Wx::AUI_NB_TOP|
      Wx::AUI_NB_TAB_SPLIT|Wx::AUI_NB_TAB_MOVE|Wx::AUI_NB_SCROLL_BUTTONS|
      Wx::AUI_NB_CLOSE_ON_ACTIVE_TAB|Wx::AUI_NB_MIDDLE_CLICK_CLOSE)
    box_sizer_2.add(@notebook, Wx::SizerFlags.new.border(Wx::ALL))

    page = Wx::Panel.new(@notebook, Wx::ID_ANY, Wx::DEFAULT_POSITION,
      Wx::DEFAULT_SIZE, Wx::TAB_TRAVERSAL)
    @notebook.add_page(page, "page #1")
    page.set_background_colour(Wx::SystemSettings.get_colour(
      Wx::SYS_COLOUR_BTNFACE))

    page_sizer = Wx::BoxSizer.new(Wx::VERTICAL)

    box_sizer = Wx::BoxSizer.new(Wx::HORIZONTAL)

    @static_text = Wx::StaticText.new(page, Wx::ID_ANY,
      "What hath Ruby wrought?")
    box_sizer.add(@static_text, Wx::SizerFlags.new.border(Wx::ALL))

    page_sizer.add(box_sizer, Wx::SizerFlags.new.border(Wx::ALL))
    page.set_sizer_and_fit(page_sizer)

    dlg_sizer.add(box_sizer_2, Wx::SizerFlags.new.border(Wx::ALL))

    set_sizer_and_fit(dlg_sizer)
    centre(Wx::BOTH)
  end
end
```

**Comment by mcorino** [comment]
Thanks for reporting this. I am not able to look into this at this moment but will pick this up when I am able to do so.

**Comment by mcorino** [comment]
I have found the root cause for this which is a `feature` (bug?) of the wxAuiManager class.
This class intercepts events for it's managed window and implements custom processing for some which in case of window destroy events causes a problem as the custom handling there blocks propagation of the event to other objects like the global App object which wxRuby NEEDS to provide reliable life cycle management for window objects.

In the case of user defined AUI manager objects for toplevel windows (dialogs or frames) wxRuby already implements a customization that fixes that problem.
Using a toplevel AUI manager for a simple dialog (or frame) with only a single Aui Notebook control is not required so the wxRuby customization is not active there. 
As there is no toplevel AUI manager active this would normally be ok as that means there is no toplevel event interception either BUT wxAuiNotebook is on of a few Aui controls that embed their own AUI manager which intercepts their events so in these cases wxRuby never gets properly notified that the wxAuiNotebook has been destroyed which causes GC segfaults as you noticed.

I have to see what the best solution here is and for that I started a discussion with the wxWidgets people as well.
